addons/base_address_city/models/res_partner.py

```diff
@@ -25,17 +25,56 @@ def _fields_view_get_address(self, arch):
         # render the partner address accordingly to address_view_id
         doc = etree.fromstring(arch)
         if doc.xpath("//field[@name='city_id']"):
-           return arch
-        for city_node in doc.xpath("//field[@name='city']"):
-            replacement_xml = """
+            return arch
+
+        replacement_xml = """
             <div>
                 <field name="country_enforce_cities" invisible="1"/>
-                <field name='city' placeholder="%s" attrs="{'invisible': [('country_enforce_cities', '=', True), ('city_id', '!=', False)], 'readonly': [('type', '=', 'contact'), ('parent_id', '!=', False)]}"/>
-                <field name='city_id' placeholder="%s" attrs="{'invisible': [('country_enforce_cities', '=', False)], 'readonly': [('type', '=', 'contact'), ('parent_id', '!=', False)]}" context="{'default_country_id': country_id}" domain="[('country_id', '=', country_id)]"/>
+                <field name='city' placeholder="%(placeholder)s" />
+                <field name='city_id' placeholder="%(placeholder)s"
+                    context="{'default_country_id': country_id}"
+                    domain="[('country_id', '=', country_id)]"/>
             </div>
-            """ % (_('City'), _('City'))
-            city_id_node = etree.fromstring(replacement_xml)
-            city_node.getparent().replace(city_node, city_id_node)
+        """ % {'placeholder': _('City')}
+
+        fields_attrs = {
+            'city': "{'invisible': [('country_enforce_cities', '=', True), ('city_id', '!=', False)], 'readonly': [('type', '=', 'contact')%s]}",
+            'city_id': "{'invisible': [('country_enforce_cities', '=', False)], 'readonly': [('type', '=', 'contact')%s]}"
+        }
+
+        def _arch_location(node):
+            in_subview = False
+            view_type = False
+            parent = node.getparent()
+            while parent is not None and (not view_type or not in_subview):
+                if parent.tag == 'field':
+                    in_subview = True
+                elif parent.tag in ['list', 'tree', 'kanban', 'form']:
+                    view_type = parent.tag
+                parent = parent.getparent()
+            return {
+                'view_type': view_type,
+                'in_subview': in_subview,
+            }
+
+        def _fill_attrs(node):
+            field_name = node.get('name')
+            new_attrs = fields_attrs.get(field_name)
+            if new_attrs:
+                location = _arch_location(node)
+                parent_condition = ''
+                if location['view_type'] == 'form' or not location['in_subview']:
+                    parent_condition = ", ('parent_id', '!=', False)"
+                replace_node.attrib['attrs'] = new_attrs % parent_condition
+
+        for city_node in doc.xpath("//field[@name='city']"):
+            replacement_tree = etree.fromstring(replacement_xml)
```

Comment:
rebuild a tree here is necessary, otherwise at the next iteration of the xpath, replacement_tree would be the same and already consumed